gtk/src/application.rs

```diff
@@ -6,11 +6,100 @@ use gio::ApplicationExt;
 use gio::ApplicationFlags;
 use glib::signal::SignalHandlerId;
 use glib::translate::*;
-use glib::ObjectExt;
+use glib::{Cast, IsA, ObjectExt, ToValue};
 
 use std::cell::RefCell;
 use std::rc::Rc;
 
+#[derive(Clone, Default)]
+pub struct ApplicationBuilder {
+    app_menu: Option<gio::MenuModel>,
+    menubar: Option<gio::MenuModel>,
+    register_session: Option<bool>,
+    action_group: Option<gio::ActionGroup>,
+    application_id: Option<String>,
+    flags: Option<gio::ApplicationFlags>,
+    inactivity_timeout: Option<u32>,
+    resource_base_path: Option<String>,
+}
+
+impl ApplicationBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn build(self) -> Result<Application, glib::BoolError> {
+        let mut properties: Vec<(&str, &dyn ToValue)> = vec![];
+        if let Some(ref app_menu) = self.app_menu {
+            properties.push(("app-menu", app_menu));
+        }
+        if let Some(ref menubar) = self.menubar {
+            properties.push(("menubar", menubar));
+        }
+        if let Some(ref register_session) = self.register_session {
+            properties.push(("register-session", register_session));
+        }
+        if let Some(ref action_group) = self.action_group {
+            properties.push(("action-group", action_group));
+        }
+        if let Some(ref application_id) = self.application_id {
+            properties.push(("application-id", application_id));
+        }
+        if let Some(ref flags) = self.flags {
+            properties.push(("flags", flags));
+        }
+        if let Some(ref inactivity_timeout) = self.inactivity_timeout {
+            properties.push(("inactivity-timeout", inactivity_timeout));
+        }
+        if let Some(ref resource_base_path) = self.resource_base_path {
+            properties.push(("resource-base-path", resource_base_path));
+        }
+        let ret = glib::Object::new::<Application>(&properties)?;
```

Comment:
What's the point of no panicking here? This can only fail if the properties or their types (and not their values!) are invalid, which can't really happen here.

Am I missing something?

Reply:
The application id could be invalid I guess?

Reply:
That causes a critical warning but `glib::Object::new()` still returns an application. It can't possibly check validity of property values here, only their types and the names of the properties.

Failing construction (because of invalid property values or similar) is only possible for objects implementing `GInitable` (see https://github.com/gtk-rs/gtk-rs/issues/137)

Reply:
Ah, great! In this case we can remove the manual call to init() from the gtk4 version and just use the "startup: signal there as well? (The signal can panic but that's just how gtk works I guess?)

And `gtk::Application::new` should just be auto generated. The `register_startup_hook` in the manual implementation should not be necessary.

Reply:
The reason why we return Result is we are calling `rt::init()?` at `build`

Reply:
@bilelmoussaoui not in gtk3. It sneakily uses `builder_postprocess = "Application::register_startup_hook(&ret);"`

Reply:
Aha, well, I guess we should use something similar in gtk4. I wasn't aware of `builder_postprocess`.

Reply:
Yeah, took me some time to figure out what's going on there.

I will take a break and suggest a new pair of patches for gtk3 and gtk4.